tools/sof-fw-boot-test.py

```diff
@@ -0,0 +1,128 @@
+
+import argparse
+import os
+import subprocess
+import sys
+import time
+
+# define command line arguments
+def parse_cmdline():
+	parser = argparse.ArgumentParser(add_help=True, formatter_class=argparse.RawTextHelpFormatter,
+	    description='A script for stress testing firmware boot')
+
+	parser.add_argument('-i', '--iter', type=int, default=100, help="number of firmware boot iterations")
+	parser.add_argument('-f', '--firmware', type=str, required=True, help='firmware filename')
+	parser.add_argument('-p', '--fw_path', type=str, required=True, help='path to the firmware file relative to /lib/firmware')
+
+	return vars(parser.parse_args())
+
+cmd_args = parse_cmdline()
+
+# Get firmware file path
+fw_path = cmd_args["fw_path"]
+full_fw_path = '/lib/firmware/' + fw_path
+if not os.path.exists(full_fw_path):
+	print("Invalid firmware path: %s" %fw_path)
+	sys.exit(1)
+
+# Get firmware file name
+fw_filename = cmd_args["firmware"]
+
+# check if the firmware binary exists
+fw_fullname = '/lib/firmware/' + fw_path + '/' + fw_filename
+if not os.path.isfile(fw_fullname):
+	print("Invalid firmware file: %s" %fw_filename)
+	sys.exit(1)
+
+num_iter = cmd_args["iter"]
+print("=======================================================================================")
+print('Starting boot stress test with:')
+print('Firmware filename: %s' %fw_filename)
+print('Path to firmware file: %s' %fw_path)
+print('Number of Iterations: %d' %num_iter)
+print("=======================================================================================")
+
+# set path to the depbugfs entries
+debugfs_path = '/sys/kernel/debug/sof/dsp_ops'
+if not os.path.exists(debugfs_path):
+	print("Debugfs path '%s' does not exist!" %debugfs_path)
+	sys.exit(1)
+
+# set the firmware filename & path
+os.system('echo %s > %s/fw_filename' %(fw_filename, debugfs_path))
+os.system('echo %s > %s/fw_path' %(fw_path, debugfs_path))
+total_boot_time = 0
+min_boot_time = sys.maxsize
+max_boot_time = 0
+
+for i in range(num_iter):
+	print("Starting firmware boot iteration: %d" %i)
+
+	# put the DSP in D3 with a timeout of 1 second
+	debugfs_entry = debugfs_path + '/dsp_power_state'
+	cmd = 'echo D3 > ' + debugfs_entry
+	subprocess.run(cmd, shell=True, check=True, timeout=1)
+
+	# check if the DSP is in D3
+	cmd = 'cat ' + debugfs_entry
+	proc = subprocess.run(cmd, shell=True, capture_output=True, check=True, text=True)
+	dsp_power_state = proc.stdout
+	dsp_power_state = dsp_power_state.rstrip()
+	if not dsp_power_state == 'D3':
+		print('Failed booting firmware. DSP is not in D3')
+		sys.exit(1)
+
+	# unload current firmware
+	debugfs_entry = debugfs_path + '/unload_fw'
+	cmd = 'echo 1 > ' + debugfs_entry
+	subprocess.run(cmd, shell=True, check=True)
+
+	# get current fw_state and continue to boot only if the current state is 'PREPARE'
+	debugfs_entry = debugfs_path + '/fw_state'
+	cmd = 'cat ' + debugfs_entry
+	proc = subprocess.run(cmd, shell=True, capture_output=True, check=True, text=True)
```

Comment:
I've replaced subprocess with python open for reading but I kept the subprocess for writing the power_state and booting so that I can use the timeout